solutions/ceoi-09-Harbingers.mdx

```diff
@@ -0,0 +1,174 @@
+---
+id: ceoi-09-Harbingers
+source: CEOI 2009
+title: Harbingers
+author: Dustin Miao
+---
+
+The official analysis can be found [here](https://studylib.net/doc/5889373/--ceoi-2009)
+
+**Time Complexity** - $\mathcal{O}(n \log n)$
+
+# Implementation 
+
+<LanguageSection>
+
+<CPPSection>
+  
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+using ll = long long;
+using ld = long double;
+using pll = pair<ll, ll>;
+
+const int MAXN = 1e5 + 1;
+
+/** A simple line class for linear functions */ 
+struct Line {
+	ll m, b;
+
+	Line(ll m = 0, ll b = 0) : m(m), b(b) {}
+
+	/** 
+	 * Evaluates the linear function at position x 
+	 */ 
+	ll operator()(ll x) { return m * x + b; }
+
+	/** 
+	 * Returns the x-coordinate of the intersection of two lines 
+	 */ 
+	friend ld intersect(Line l1, Line l2) { 
+		return (ld)(l2.b - l1.b) / (ld)(l1.m - l2.m); 
+	}
+};
+
+int N;
+ll S[MAXN];  // Start time for harbinger at node
+ll V[MAXN];  // Velocity for harbinger at node
+ll min_time[MAXN];  // Minimum time to reach the capital from node
+
+vector<pll> T[MAXN];  // Stores tree edges
+
+Line stk[MAXN];  // convex hull of lines
+int stk_max = 0;  // end of convex hull stack
+
+/**
+ * Given the convex hull of points in stk[],
+ * finds the minimum y value for the given x value
+ * out of all lines in the hull
+ * @param x The x position
+ * @return The minimum y value in the convex hull
+ */ 
+ll line_min(ll x) {
+	// binary search for min value
+	int l = 0, r = stk_max - 1;
+	while (l < r) {
+		int m = (l + r) / 2;
+		if (intersect(stk[m], stk[m + 1]) < x) {
+			l = m + 1;
+		} else {
+			r = m;
+		}
+	}
+	return stk[l](x);
+}
+
+/**
+ * Gives the position such that if this line was to be added 
+ * into the convex hull, it would occupy that position. The 
+ * convex hull is stored in stk. If no lines are to be 
+ * removed to make room for the new line, then the size of the
+ * convex hull is returned. Note that the convex hull 
+ * is NOT modified in this function. 
+ * @param line The line to be queried
+ * @return An index between 1 and stk_max (inclusive) specifying 
+ *		 the location the given line would occupy should it 
+ *		 be added to the hull
+ */ 
```

Comment:
```suggestion
```

Explain this in editorial if needed

Reply:
but all of this stuff is implementation specific?

Reply:
implementation-specific editorial. this much commenting is not needed

Reply:
implementation-specific editorial. this much commenting is not needed

Reply:
It's better than nothing? how is more (useful) comments a bad thing? also, the editorial that is linked is pretty good, so there's no need to reiterate redundant information